File: src/main/java/me/helloc/iot/tunnel/MessageBuffer.java
package me.helloc.iot.tunnel;

import java.util.Map;

/**
 * Buffer storage for messages delivered via MQTT.
 * Implementations may store messages in memory or external systems.
 */
public interface MessageBuffer {
    /**
     * Adds a message for the given topic to the buffer.
     */
    void add(String topic, String message);

    /**
     * Retrieves and removes the earliest buffered message or returns {@code null} if empty.
     */
    Map.Entry<String, String> poll();
}

File: src/main/kotlin/me/helloc/iot/tunnel/MessageBuffer.kt
package me.helloc.iot.tunnel

/**
 * Buffer storage for messages delivered via MQTT.
 * Implementations may store messages in memory or external systems.
 */
interface MessageBuffer {
    /**
     * Adds a message for the given topic to the buffer.
     */
    fun add(topic: String, message: String)

    /**
     * Retrieves and removes the earliest buffered message or returns `null` if empty.
     */
    fun poll(): Pair<String, String>?
}
